File: smartlist/src/main/java/com/smartlist/module/candidate/questionstrategy/SurveyStrategy.java
package com.smartlist.module.candidate.questionstrategy;

import com.smartlist.model.Question;
import com.smartlist.module.candidate.CandidateResponseCodes;
import com.smartlist.module.candidate.dto.request.SubmitAnswerReqDTO;
import com.smartlist.module.question.AnswerOptionRepository;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class SurveyStrategy extends BaseQuestionStrategy{

    public SurveyStrategy(final SubmitAnswerReqDTO submitAnswerReqDto, final Question question,
                          final AnswerOptionRepository answerOptionRepository){
        super(submitAnswerReqDto, question, answerOptionRepository);

        if(this.submitAnswerReqDto.getAnswerIds().isEmpty()){
            throw CandidateResponseCodes.INVALID_ANSWER_ID;
        }
    }

    @Override
    public int score() {
        List<Integer> answerOptions = this.submitAnswerReqDto.getAnswerIds();
        List<Integer> correctAnswerOptions = new ArrayList<>();
        this.question.getAnswerOptions().forEach(answerOption -> {
            if (answerOption.isCorrect()) {
                correctAnswerOptions.add(answerOption.getId());
            }
        });
        correctAnswerOptions.sort(Comparator.naturalOrder());
        answerOptions.sort(Comparator.naturalOrder());
        int matchCount = 0;

        for (Integer correctAnswerOption : correctAnswerOptions) {
            if (answerOptions.contains(correctAnswerOption)) {
                matchCount++;
            }
        }

        return matchCount > 0 ? this.question.getScore() : 0;

    }
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/candidate/candidate-question-strategy/survey-strategy.ts
import { Injectable } from '@nestjs/common';
import { Question } from 'src/entities/question.entity';
import { SubmitAnswerReqDto } from 'src/modules/candidate/dto/request/submit-answer.req.dto';
import { BaseQuestionStrategy } from './base-question-strategy';

@Injectable()
export class SurveyStrategy extends BaseQuestionStrategy {
  constructor(submitAnswerReqDto: SubmitAnswerReqDto, question: Question) {
    super(submitAnswerReqDto, question);
  }

  score() {
    const answerOptions: number[] = this.submitAnswerReqDto.answer_ids;
    const correctAnswerOptions: number[] = [];
    this.question.answerOptions.forEach((answerOption) => {
      if (answerOption.correct) {
        correctAnswerOptions.push(answerOption.id);
      }
    });
    correctAnswerOptions.sort();
    answerOptions.sort();
    let matchCount = 0;

    for (let i = 0; i < correctAnswerOptions.length; i++) {
      if (answerOptions.indexOf(correctAnswerOptions[i]) != -1) {
        matchCount++;
      }
    }

    return matchCount > 0 ? this.question.score : 0;
  }
}
